src/controllers/state_controller.hpp

```diff
@@ -0,0 +1,1101 @@
+    else // if (device_ == Device::GPU)
+#ifdef AER_THRUST_SUPPORTED
+      if (precision_ == Precision::Double)
+        state_ = std::make_shared<DensityMatrix::State<QV::DensityMatrixThrust<double>>>();
+      else
+        state_ = std::make_shared<DensityMatrix::State<QV::DensityMatrixThrust<float>>>();
+#else
+      throw std::runtime_error("specified method does not support non-CPU device: method=density_matrix");
+#endif
+  } else if (method_ == Method::unitary) {
+    if (device_ == Device::CPU)
+      if (precision_ == Precision::Double)
+        state_ = std::make_shared<QubitUnitary::State<QV::UnitaryMatrix<double>>>();
+      else
+        state_ = std::make_shared<QubitUnitary::State<QV::UnitaryMatrix<float>>>();
+    else // if (device_ == Device::GPU)
+#ifdef AER_THRUST_SUPPORTED
+      if (precision_ == Precision::Double)
+        state_ = std::make_shared<QubitUnitary::State<QV::UnitaryMatrixThrust<double>>>();
+      else
+        state_ = std::make_shared<QubitUnitary::State<QV::UnitaryMatrixThrust<float>>>();
+#else
+      throw std::runtime_error("specified method does not support non-CPU device: method=unitary");
+#endif
+  } else if (method_ == Method::matrix_product_state) {
+    if (device_ == Device::CPU)
+      state_ = std::make_shared<MatrixProductState::State>();
+    else // if (device_ == Device::GPU)
+        throw std::runtime_error("specified method does not support non-CPU device: method=matrix_product_state");
+  } else if (method_ == Method::stabilizer) {
+    if (device_ == Device::CPU)
+      state_ = std::make_shared<Stabilizer::State>();
+    else // if (device_ == Device::GPU)
+        throw std::runtime_error("specified method does not support non-CPU device: method=stabilizer");
+  } else if (method_ == Method::extended_stabilizer) {
+    if (device_ == Device::CPU)
+      state_ = std::make_shared<ExtendedStabilizer::State>();
+    else // if (device_ == Device::GPU)
+        throw std::runtime_error("specified method does not support non-CPU device: method=extended_stabilizer");
+  } else if (method_ == Method::superop) {
+    if (device_ == Device::CPU)
+      if (precision_ == Precision::Double)
+        state_ = std::make_shared<QubitSuperoperator::State<QV::Superoperator<double>>>();
+      else
+        state_ = std::make_shared<QubitSuperoperator::State<QV::Superoperator<float>>>();
+    else // if (device_ == Device::GPU)
+        throw std::runtime_error("specified method does not support non-CPU device: method=superop");
+  } else {
+      throw std::runtime_error("not supported method.");
+  }
+
+#ifdef _OPENMP
+  if (parallel_state_update_ == 0) {
+    parallel_state_update_ = omp_get_max_threads();
+  }
+#endif
+  state_->set_config(configs_);
+  state_->set_distribution(1);
+  state_->set_max_matrix_qubits(max_gate_qubits_);
+  state_->set_parallelization(parallel_state_update_);
+
+  state_->initialize_qreg(num_of_qubits_);
+  state_->initialize_creg(num_of_qubits_, num_of_qubits_);
```

Comment:
I think number of classical registers and memories sometimes maybe larger than number of qubits. These should be modified by option.